In [84]:
import math
import random

def generate_data(num_trials):
    trials = []
    for _ in range(num_trials):
        count = 0
        while True:
            count += 1
            if random.randint(1, 125) == 1:
                break
        trials.append(count)
    return trials

x_data = generate_data(5000)

def apply_formula(x):
    # 여기에 당신의 공식을 적용합니다.
    # 예: y = x * 2
    y = round(1-math.exp(-x/125),4)
    return y

def generate_y_list(x_list):
    y_list = []
    for x in x_list:
        y = apply_formula(x)
        y_list.append(y)
    return y_list

# Y 데이터 리스트 생성
y_data = generate_y_list(x_data)

print("X 데이터:", x_data)
#print("Y 데이터:", y_data)

X 데이터: [106, 260, 73, 450, 55, 223, 108, 123, 106, 12, 73, 530, 94, 74, 29, 3, 435, 72, 171, 96, 314, 490, 141, 6, 33, 228, 8, 29, 18, 7, 22, 219, 106, 71, 332, 78, 54, 57, 141, 77, 96, 261, 93, 102, 222, 350, 135, 137, 153, 38, 6, 18, 185, 189, 241, 11, 164, 434, 69, 49, 21, 91, 112, 94, 10, 31, 102, 2, 270, 152, 19, 10, 126, 17, 37, 228, 67, 88, 37, 38, 105, 43, 5, 305, 228, 91, 5, 33, 137, 143, 36, 124, 64, 546, 454, 34, 2, 226, 375, 294, 60, 91, 56, 77, 25, 186, 249, 97, 192, 104, 129, 243, 213, 40, 96, 132, 273, 51, 66, 179, 224, 327, 16, 61, 137, 101, 148, 17, 127, 167, 185, 41, 1, 113, 38, 232, 138, 45, 90, 188, 169, 178, 89, 36, 377, 208, 29, 240, 76, 30, 286, 125, 124, 73, 25, 162, 535, 42, 21, 440, 28, 106, 51, 35, 4, 23, 116, 44, 97, 196, 563, 41, 122, 161, 73, 31, 114, 150, 27, 97, 32, 125, 24, 81, 85, 41, 171, 105, 19, 215, 107, 18, 148, 32, 109, 177, 482, 50, 114, 73, 161, 225, 218, 186, 141, 33, 40, 194, 91, 178, 55, 163, 71, 97, 225, 151, 25, 15, 173, 144, 80, 36, 6, 75

In [85]:
# 데이터셋 생성 함수
def create_dataset(numbers, n_steps):
    X, y = [], []
    for i in range(len(numbers)):
        end_ix = i + n_steps
        if end_ix > len(numbers)-1:
            break
        seq_x, seq_y = numbers[i:end_ix], numbers[end_ix] # end_ix는 포함되지 않는다.
        X.append(seq_x)
        y.append(seq_y)
    return np.array(X), np.array(y)

# 숫자 시퀀스 정의
numbers = x_data
n_steps = 5 # 학습하고싶은 숫자묶음의 크기(발걸음)을 설정합니다.

# 데이터셋 생성
X, y = create_dataset(numbers, n_steps) #데이터셋 생성 함수 활성화

print(X)
print(y)

[[106 260  73 450  55]
 [260  73 450  55 223]
 [ 73 450  55 223 108]
 ...
 [316  11 305  19  27]
 [ 11 305  19  27   6]
 [305  19  27   6 139]]
[223 108 123 ...   6 139  10]


In [86]:
!pip install GPy

In [87]:
import numpy as np
from sklearn.gaussian_process import GaussianProcessRegressor
from sklearn.gaussian_process.kernels import Matern, WhiteKernel

# 매턴 커널과 화이트 커널(노이즈 항)을 조합하여 커널 정의
kernel = Matern(length_scale=1.0, nu=1.5) + WhiteKernel(noise_level=1)

# 가우시안 프로세스 회귀 모델 생성
gp = GaussianProcessRegressor(kernel=kernel)

# 가우시안 프로세스 데이터 예시1

- 입력 데이터
X_train = np.array([[1], [2], [3], [4], [5], [6], [7], [8], [9], [10]])

- 타겟 데이터 (y = x^2 + 3x + 2)
y_train = np.array([6, 12, 20, 30, 42, 56, 72, 90, 110, 132])

# 가우시안 프로세스 데이터 예시2

-입력데이터
[1]이 아니고 [1,3,6]과 같이 여러개일때는 자동차의 연비, 무게, 생산년원과같이 특성이 여러개인 경우에 해당한다.

In [88]:
# 모델 학습
gp.fit(X, y)

 /Users/pc/anaconda3/lib/python3.10/site-packages/sklearn/gaussian_process/kernels.py:430: ConvergenceWarning:The optimal value found for dimension 0 of parameter k1__length_scale is close to the specified upper bound 100000.0. Increasing the bound and calling fit again may find a better value.


GaussianProcessRegressor(kernel=Matern(length_scale=1, nu=1.5) + WhiteKernel(noise_level=1))

In [90]:
# 새로운 데이터에 대한 예측
X_new = np.array([[35, 86, 162, 6, 31]])
y_pred, sigma = gp.predict(X_new, return_std=True)

# 예측 결과와 불확실성 출력
print("예측된 값:", y_pred)
print("예측 불확실성:", sigma)

예측된 값: [19.9265098]
예측 불확실성: [162.70387131]


# 두번째 gaussian process 모델 생성